spec/controllers/books_controller_spec.rb

```diff
@@ -0,0 +1,144 @@
+require 'rails_helper'
+
+RSpec.describe BooksController, type: :controller do
+  describe 'GET /index' do
+    subject { get :index }
+
+    it 'returns all books' do
+      expect(subject.status).to eq(200)
+      expect(assigns(:books)).to eq(Book.all)
+    end
+  end
+
+  describe 'GET /show' do
+    subject { get :show, params: }
+
+    let(:book) { create(:book) }
+    let(:params) { { id: book.id } }
+
+    context 'when book exist' do
+      it 'returns book' do
+        expect(subject.status).to eq(200)
+        expect(subject).to render_template(:show)
+        expect(assigns(:book)).to eq(Book.find(book.id))
+      end
+    end
+
+    context 'when book not exist' do
+      let(:params) { { id: -1 } }
+
+      it 'raise error not found' do
+        # expect(subject.status).to eq(404)
+        expect { subject }.to raise_error(ActiveRecord::RecordNotFound)
+      end
+    end
+  end
+
+  describe 'GET /new' do
+    subject { get :new }
+
+    it 'renders to new book form' do
+      expect(subject.status).to eq(200)
+      expect(subject).to render_template(:new)
+    end
+  end
+
+  describe 'POST /create' do
+    subject { post :create, params: }
+
+    let(:params) { { book: { name: 'Controller book', description: 'Book from controller', release: '2023-01-01' } } }
+
+    it 'creates a book success' do
+      expect(subject.status).to eq(302)
+      expect(Book.count).to eq(1)
+
+      # test value of each attributes
+      created_book = assigns(:book)
+      expect(Book.find(created_book.id).name).to eq('Controller book')
+      expect(Book.find(created_book.id).description).to eq('Book from controller')
+      expect(Book.find(created_book.id).release).to eq(Date.new(2023, 1, 1))
+
+      # redirect to show page
+      expect(subject).to redirect_to(book_path(created_book.id))
+    end
+
+    context 'validation' do
+      let(:params) { { book: { name: '' } } }
+
+      it 'requires name' do
+        expect(subject.status).to eq(200)
+        expect(assigns(:book)).not_to be_valid
+        expect(Book.count).to eq(0)
+
+        # render new page
+        expect(subject).to render_template(:new)
+      end
+    end
+  end
+
+  describe 'GET /edit' do
+    subject { get :edit, params: }
+
+    let(:book) { create(:book) }
+    let(:params) { { id: book.id } }
+
+    it 'renders to edit book form' do
+      expect(subject.status).to eq(200)
+      expect(subject).to render_template(:edit)
+    end
+  end
+
+  describe 'PUT /update' do
+    subject { put :update, params: }
+
+    let!(:book) { create(:book) }
+    let(:invalid_attributes) { { name: '', description: 'invalid name' } }
+
+    it 'renders the edit page for invalid attributes' do
+      put :update, params: { id: book.id, book: invalid_attributes }
```

Comment:
can we try
```suggestion
      params = { id: book.id, book: invalid_attributes }
```
and call subject as always

Reply:
พอลองตามโค้ดด้านล่างมันได้ status เป็น 302 แทนอะค่ะ ไม่มั่นใจว่าทำไม
```
params = { id: book.id, book: invalid_attributes }
subject
expect(subject.status).to eq(200)
```
อันนี้โค้ด book_controller
```
def update
    if @book.update(book_params)
      redirect_to @book
    else
      render :edit
    end
  end
```

Reply:
งั้นเอา context มาครอบแล้วใช้ let(:params) เหมือนเดิมก็ได้ครับ

```
context 'when updates with invalid paramter' do
  let(:params) { ... }

  it 'renders the edit page' do
    ...
  end
end

context 'when updates with valid paramter' do
  let(:params) { ... }

  it 'updates book and redirect to show page' do
   ...
  end
end

```